src/passes/OptimizeInstructions.cpp

```diff
@@ -841,6 +757,54 @@ struct OptimizeInstructions
           }
         }
       }
+      if (auto* constTrue = select->ifTrue->dynCast<Const>()) {
+        if (auto* constFalse = select->ifFalse->dynCast<Const>()) {
+          if (select->type == Type::i32 || select->type == Type::i64) {
+            auto trueValue = constTrue->value.getInteger();
+            auto falseValue = constFalse->value.getInteger();
+            if ((trueValue == 1LL && falseValue == 0LL) ||
+                (trueValue == 0LL && falseValue == 1LL)) {
+              Builder builder(*getModule());
+              // canonicalize "expr ? 0 : 1" to "!expr ? 1 : 0"
+              if (trueValue == 0LL) {
+                select->condition = optimizeBoolean(
+                  builder.makeUnary(EqZInt32, select->condition));
+              }
+              Expression* condition = nullptr;
+              if (Properties::emitsBoolean(select->condition)) {
+                // !x ? 1 : 0   ==>   !x
+                // x <=> y ? 1 : 0   ==>   x <=> y
+                condition = select->condition;
+              } else {
+                // expr ? 1 : 0   ==>   !!expr
+                if (getPassOptions().shrinkLevel > 0) {
+                  condition = builder.makeUnary(
+                    EqZInt32, builder.makeUnary(EqZInt32, select->condition));
+                } else {
+                  condition =
+                    builder.makeBinary(NeInt32,
+                                       select->condition,
+                                       builder.makeConst(Literal(int32_t(0))));
+                }
+              }
+              return select->type == Type::i64
+                       ? builder.makeUnary(ExtendUInt32, condition)
+                       : condition;
+            }
+          }
+        }
+      }
+      if (auto* condition = select->condition->dynCast<Unary>()) {
+        if (condition->op == EqZInt32) {
+          // flip select to remove eqz, if we can reorder
+          EffectAnalyzer ifTrue(getPassOptions(), features, select->ifTrue);
+          EffectAnalyzer ifFalse(getPassOptions(), features, select->ifFalse);
+          if (!ifTrue.invalidates(ifFalse)) {
```

Comment:
this can be `canReorder` as well, which is shorter.

Reply:
this part is just moved from [here](https://github.com/WebAssembly/binaryen/pull/2869/files/0706ee8e55930169a5923260a22796d787bea103#diff-54e02686df3480c98e3f90a24705c00bL812) and don't touched so I prefer keep it as original.

Reply:
Ok, I'll clean this up separately then.